### PR TITLE
Document expectations for usage of new JS features

### LINKED ISSUE #6266: Formalize policy for usage of new language features
The usage of new ECMAScript features has been a topic of discussion in a few recent pull requests. Currently, there does not seem to be consensus around in what contexts it is appropriate to use such features in WPT. Having clarity here would help everyone involved with this project: contributors (so they could have some confidence that their stylistic choices would be accepted), reviewers (so they wouldn't have to resolve discrepancies between the preferences of different contributors), and consumers (so they would have a realistic expectation about when they might start to fail tests due to orthogonal issues in their JS runtime).

Unfortunately, the desires of those audiences is somewhat at odds. Test contributors will tend to want to use all the tools that are available to them. On the other hand, implementers will generally prefer to avoid practices that could prevent them from executing tests. (For their part, maintainers probably just want clarity--the specific decision doesn't effect them as much.)

This tension might be partially addressed by **a "contextual" policy**. Maybe new ECMAScript features are acceptable in tests written against specifications that are similarly new. Along these lines, new features would also be barred from shared utilities such as those found in the `common/` and `resources/` top-level directories (since those files may be included by any test and would therefore be a lowest-common-denominator). Defining, documenting, and enforcing this kind of policy is a little tricky, though, because it assumes familiarity with the publication history of various specifications.

Under **a "blanket" policy**, WPT would simply declare, "any test may use any language feature that was standardized at least X years prior to the current date." For example, setting up a blanket policy of 1 year would today translate to: "you may use the exponentiation operator, but you may not use async functions." This would be easier to follow in practice, although it might also needlessly shut some consumers out from tests that would otherwise help them ensure conformance.

I would love to get input on this topic from members of all three audiences. My goal is to amend [`docs/_writing-tests/assumptions.md`](https://github.com/w3c/web-platform-tests/blob/master/docs/_writing-tests/assumptions.md) and/or [`docs/_writing-tests/general-guidelines.md`](https://github.com/w3c/web-platform-tests/blob/master/docs/_writing-tests/general-guidelines.md) with a policy by Friday June 23.

### PR DESCRIPTION
**Pull request by jugglinmike**
These expectations were discussed by project maintainers at:

https://github.com/w3c/web-platform-tests/issues/6266

This is intended to resolve gh-6266.

### CODE REVIEW

**Comment by tobie** [comment]
LGTM

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246387523)

# Firefox (nightly) #
Testing web-platform-tests at revision 8b77d6bad8fb2e9d906b527a488473cfc6c9c361
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246387524)

# Sauce (safari) #
Testing web-platform-tests at revision 8b77d6bad8fb2e9d906b527a488473cfc6c9c361
Using browser at version 10.0
Starting 10 test iterations

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246387530)

# Chrome (unstable) #
Testing web-platform-tests at revision 8b77d6bad8fb2e9d906b527a488473cfc6c9c361

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246387531)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 8b77d6bad8fb2e9d906b527a488473cfc6c9c361

**Review by sideshowbarker — COMMENTED** [review]
LGTM but I think @gsnedders should be the one to OK this

**Comment by gsnedders** [comment]
Two thoughts:

 * This should be under "Be cross-platform" given that talks about what can assumed.
 * It should probably be more generic—the policy isn't specific to JS features, but rather features in general.

**Comment by jugglinmike** [comment]
You got it. I moved the text to a new sub-section of "Be Cross Platform" named
"Be Conservative". I also included a reference to https://caniuse.com as this
is more helpful for determining browser support of non-ECMAScript features.

This change called for a modification of the commit message, so I've
force-pushed a new commit to this branch. The original commit is available at:
https://github.com/bocoup/web-platform-tests/tree/document-language-features-orig

**Review by tobie — COMMENTED** [review]
So I'm not sure how much has change between the two commits, but I'm really uneasy with the tone here. It sounds paternalistic, tbh.

Using modern features is sort of described as being lazy and we should aim to "be conservative" instead. I don't think I can stand behind this depiction.

I think this points to a deeper lack in consensus as to which direction (and underlying goals and purpose) this project should hold.

Are we trying to support IE6, or bleeding edge browsers?

Are we there to help implementors implement or sites like caniuse, test?

This middle-ground leaves me uneasy. We should either embrace modern JS or backwards compat, but not a little bit of both.

**Comment by tobie** [comment]
Note I don't think your effort to write this down is poor or in vain. Quite the contrary, actually. It reflects the lack of consensus around major project goals. It's great we're having this conversation. So thank you!

**Comment by jgraham** [comment]
I think only `testharness.js` (and maybe related things like `idlharness.js`) is really controversial here. What I personally care about for those is the usual criteria (Chrome, Edge Firefox, Safari) + works in Servo. I am prepared to enforce that level of support.

However there have been cases where people have asked for changes to support older browsers, and where they have been small we have usually accepted those changes. So empirically some people care about that, and I don't think it's worth gratuitously breaking them. But I would certainly consider any such uses to be "Tier 2", by which i mean, worth taking patches for and bearing in mind, but not worth blocking features on.

File: docs/_writing-tests/general-guidelines.md
Comment (jugglinmike) [inline docs/_writing-tests/general-guidelines.md]: I was trying to communicate the severity of the failures--that they will span
multiple tests across all of WPT. But that detail is not particularly important
for this document, and the statement is no less accurate without the word, so
I'm happy to remove it.

File: docs/_writing-tests/general-guidelines.md
Comment (jugglinmike) [inline docs/_writing-tests/general-guidelines.md]: That phrase only describes features that would *not* be acceptable according to
the criteria we've identified here. This text needs to apply to *any* feature
since the explanation that follows describes how to distinguish between
"acceptable" and "unacceptable." I chose the words "unrelated" and "extraneous"
to differentiate the features under test (whose validity is unquestionable)
from those that are there to facilitate the test (whose validity is determined
by the criteria documented here).

To try make this more clear, I've rewritten that sentence:

> it may be convenient to use some other feature (e.g. a web browser API or
> ECMAScript language feature) that is not directly related to the test itself

**Comment by jugglinmike** [comment]
> So I'm not sure how much has change between the two commits, but I'm really
> uneasy with the tone here. It sounds paternalistic, tbh.

It sounds like my personal bias leaked into the copy. I think paternalistic is
a somewhat negative interpretation, but I certainly didn't mean to offend.
While I'm happy to substitute in some more neutral words, this is a more
subjective change, and I'd like to avoid potential back-and-forth in satisfying
the expectations of multiple reviewers along these lines. That's why I'm
requesting a second opinion on the tone: @gsnedders, do you agree with
@tobie's suggestion?

**Comment by tobie** [comment]
> I think paternalistic is a somewhat negative interpretation.

Yeah, I struggled to find a more toned down word but couldn't. It was late and I went with that. Sorry. :-/

**Comment by jugglinmike** [comment]
No worries!

File: docs/_writing-tests/general-guidelines.md
Comment (gsnedders) [inline docs/_writing-tests/general-guidelines.md]: This seems to be duplicating what is under "Be Minimal" above. (Though "Be Conservative" is probably a better header!) We should probably add a paragraph there rather than this.

My suggestion would be something like:

> Tests should ensure they run correctly and fail when appropriate (including, e.g., when the feature is unsupported) on the latest stable release of Apple Safari, Google Chrome, Microsoft Edge, and Mozilla Firefox. They should, therefore, not rely on any features aside from the one under test unless they are supported in all four browsers.
>
> Patches to make tests run on older versions or other browsers will be accepted provided they are relatively simple and do not add undue complexity to the test.

File: docs/_writing-tests/general-guidelines.md
Comment (gsnedders) [inline docs/_writing-tests/general-guidelines.md]: Let's just keep the `testharness.js` exception (and I think it is the only exception) documented in the header of it, and not have it here. I think it's sufficiently obscure that it needn't be here.

**Review by gsnedders — CHANGES_REQUESTED** [review]

**Comment by jugglinmike** [comment]
@gsnedders Thanks! I've removed that new section, renamed the existing one, and
added your test. I modified your suggestion a bit by being explicit about "run
correctly and fail when appropriate":

> Tests should pass when the feature under test exposes the expected behavior,
> and they should fail when the feature under test is not implemented or is
> implemented incorrectly.

And as discussed on IRC, I've persisted the references to shared resources and
to the community-run projects that track implementation status.

File: docs/_writing-tests/general-guidelines.md
Comment (tobie) [inline docs/_writing-tests/general-guidelines.md]: Should we talk about rendering engines, here, instead of browsers?

**Comment by snuggs** [comment]
Fixes #6266

File: docs/_writing-tests/general-guidelines.md
Comment (zcorpan) [inline docs/_writing-tests/general-guidelines.md]: May be worth encouraging that when a contributor runs into an unrelated feature not working as expected in the 4 browsers, that they write a separate test for that, or at least report an issue about it so somebody else can follow up.

File: docs/_writing-tests/general-guidelines.md
Comment (foolip) [reply docs/_writing-tests/general-guidelines.md]: Rendering engines don't have releases, so that wouldn't quite work.

File: docs/_writing-tests/general-guidelines.md
Comment (foolip) [inline docs/_writing-tests/general-guidelines.md]: @jugglinmike, depending on Promises is fine now, but was introduced in ECMAScript 2015.

I suggest dropping this paragraph, and pointing to https://wpt.fyi/results/?label=stable in the previous. No specific guidance we give will stay up to date.

**Review by foolip — APPROVED** [review]
LGTM % nits

**Comment by jugglinmike** [comment]
> I suggest dropping this paragraph,

The policy [was](https://github.com/web-platform-tests/wpt/issues/6266#issuecomment-309444814) and [continues to be](https://github.com/web-platform-tests/wpt/pull/12399#issuecomment-412470957) different for shared infrastructure. We can remove this paragraph, but we'll need to replace it with something that documents the distinction.

> and pointing to https://wpt.fyi/results/?label=stable in the previous.

The preceding paragraph describes ECMAScript conformance, but wpt.fyi doesn't help people understand that. To make this more clear, I've changed "For features that are not used in existing tests" to "For language features that are not used in existing tests"

> No specific guidance we give will stay up to date.

Given that WPT's policy is partially based on Servo's implementation status and that [Servo's conformance is not publicly visible](https://github.com/web-platform-tests/wpt/issues/6266#issuecomment-309547561), I think we're forced to maintain a static/"rot-prone" description in this document. I've updated the patch to reference ES2017 in recognition of Servo's lack of support for async functions.

**Comment by gsnedders** [comment]
@jugglinmike I *think* Servo's support should basically match FF50, which es-compat at least used to have… :/

**Comment by jugglinmike** [comment]
Rebased again. The prior version of this branch is available here:

https://github.com/bocoup/wpt/tree/document-language-features-2

**Comment by gsnedders** [comment]
Looking over this again, this looks broadly good to me?

**Review by gsnedders — APPROVED** [review]

**Comment by sideshowbarker** [comment]
Can we merge this? (After rebasing and fixing the conflicts.)

File: docs/_writing-tests/general-guidelines.md
Comment (stephenmcgruer) [inline docs/_writing-tests/general-guidelines.md:128]: This hasn't aged well, imo. The only thing in es2017 I'm aware of that isn't well supported across all major browsers (outside of IE11) is `SharedArrayBuffer`. More importantly, `async` functions are in es2017 and we certainly have a lot of those!

I think this would need to change to at least 2018 if not 2019, and moreso I'm concerned with us forgetting to update this every few years as features become widely available.

Comment (sideshowbarker) [reply docs/_writing-tests/general-guidelines.md:128]:
```suggestion
are relatively new and not yet supported in all current browser engines should not be used. This
```

**Comment by stephenmcgruer** [comment]
Again, jugglinmike no longer works in the WPT space afaik, so I've uploaded a new version of this at https://github.com/web-platform-tests/wpt/pull/26204 which incorporates the concern about the es2017 line. Hopefully we can get that merged.

**Pull request closed without merging**
